ai_eval/llm_services.py

```diff
@@ -81,25 +117,56 @@ def _get_headers(self):
         self._ensure_token()
         return {'Authorization': f'Bearer {self._access_token}'}
 
-    def get_response(self, model, api_key, messages, api_base):
+    def get_response(
+            self,
+            model,
+            api_key,
+            messages,
+            api_base,
+            thread_id=None,
+            use_threads=False
+    ):
         """
         Send completion request to custom LLM endpoint.
         """
         url = self.completions_url
+        # When reusing an existing thread, only send the latest user input and rely on
+        # the provider to apply prior context associated with the conversation_id.
+        if use_threads and thread_id:
+            latest_user = None
+            for msg in reversed(messages):
+                if (msg.get('role') or '').lower() == 'user':
+                    latest_user = msg.get('content', '').strip()
+                    break
+            prompt = f"User: {latest_user}" if latest_user is not None else ""
+        else:
+            prompt = " ".join(
+                f"{msg.get('role', '').capitalize()}: {msg.get('content', '').strip()}"
+                for msg in messages
+            )
         # Adjust the payload structure based on custom API requirements
-        prompt = " ".join(
-            f"{msg.get('role', '').capitalize()}: {msg.get('content', '').strip()}"
-            for msg in messages
-        )
         payload = {
             "model": str(model),
             "prompt": prompt,
         }
+        # If threads are enabled, pass through conversation id when available
+        # and let the provider initialize a new thread if not.
+        if use_threads and thread_id:
+            payload["conversation_id"] = thread_id
+
         response = requests.post(url, json=payload, headers=self._get_headers(), timeout=10)
         response.raise_for_status()
         data = response.json()
         # Adjust this if custom API returns the response differently
-        return data.get("response")
+        text = data.get("response")
+        # Try to capture a new conversation/thread id if the API returns one.
+        new_thread_id = None
+        if use_threads:
+            new_thread_id = (
+                data.get("conversation_id")
+                or (data.get("data", {}) if isinstance(data.get("data"), dict) else {}).get("conversation_id")
+            )
```

Comment:
I think this code is a bit hard to follow, it would be better to expand it a bit to make it easier to follow:


```suggestion
        if use_threads:
            new_thread_id = data.get("conversation_id")
            if not new_thread_id and isinstance(data.get("data"), dict):
                new_thread_id = data["data"].get("conversation_id")
```

Reply:
Updated.